soh/soh/Enhancements/crowd-control/CrowdControl.cpp

```diff
@@ -260,328 +270,189 @@ CrowdControl::Effect* CrowdControl::ParseMessage(char payload[512]) {
-            if ((gSaveContext.health - (16 * value)) <= 0) {
-                return EffectResult::Failure;
-            }
-            
-            if (dryRun == 0) CMD_EXECUTE(fmt::format("{} {}", effectId, value));
-            return EffectResult::Success;
-        }
+    if (effect->category == "") {
+        effect->category = "none";
     }
 
-    return EffectResult::Retry;
+    return effect;
 }
 
-bool CrowdControl::SpawnEnemy(std::string effectId) {
-    Player* player = GET_PLAYER(gPlayState);
-
-    int enemyId = 0;
-    int enemyParams = 0;
-    float posXOffset = 0;
-    float posYOffset = 0;
-    float posZOffset = 0;
-
-    if (effectId == EFFECT_SPAWN_WALLMASTER) {
-        enemyId = 17;
-    } else if (effectId == EFFECT_SPAWN_ARWING) {
-        // Don't allow Arwings in certain areas because they cause issues.
-        // Locations: King dodongo room, Morpha room, Twinrova room, Ganondorf room, Fishing pond, Ganon's room
-        // TODO: Swap this to disabling the option in CC options menu instead.
-        if (gPlayState->sceneNum == SCENE_DDAN_BOSS || gPlayState->sceneNum == SCENE_MIZUSIN_BS ||
-            gPlayState->sceneNum == SCENE_JYASINBOSS || gPlayState->sceneNum == SCENE_GANON_BOSS ||
-            gPlayState->sceneNum == SCENE_TURIBORI || gPlayState->sceneNum == SCENE_GANON_DEMO) {
-            return 0;
-        }
-        enemyId = 315;
-        enemyParams = 1;
-        posYOffset = 100;
-    } else if (effectId == EFFECT_SPAWN_DARK_LINK) {
-        enemyId = 51;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_STALFOS) {
-        enemyId = 2;
-        enemyParams = 2;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_WOLFOS) {
-        enemyId = 431;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_FREEZARD) {
-        enemyId = 289;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_KEESE) {
-        enemyId = 19;
-        enemyParams = 2;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_ICE_KEESE) {
-        enemyId = 19;
-        enemyParams = 4;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_FIRE_KEESE) {
-        enemyId = 19;
-        enemyParams = 1;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_TEKTITE) {
-        enemyId = 27;
-        posXOffset = 75;
-        posYOffset = 50;
-    } else if (effectId == EFFECT_SPAWN_LIKE_LIKE) {
-        enemyId = 221;
-        posXOffset = 75;
-        posYOffset = 50;
-    }
+CrowdControl::EffectResult CrowdControl::CanApplyEffect(Effect* effect) {
+    // Get result for any given effect, and transform that into CC's own enums.
+    EffectResult result;
+    GameInteractionEffectQueryResult queryResult = GameInteractor::CanApplyEffect(effect->giEffect);
```

Comment:
seeing this i realize my earlier comments were probably oversimplifying. I still stand by the "we should simplify the logic by having the apply effect methods do the `CanApply` check" idea, but it seems we'd also need to add a little translation method that takes in a `GameInteractionEffectQueryResult` and spits out a `CrowdControl::EffectResult`